Pod/Classes/GraphQL/ARGraphQLQueryCache.m

```diff
@@ -34,63 +34,68 @@ + (void)load;
-        dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
-            NSURL *preHeatedCache = [[NSBundle mainBundle] URLForResource:@"PreHeatedGraphQLCache" withExtension:nil];
-            NSError *error = nil;
-            NSArray<NSURL *> *cacheEntries = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:preHeatedCache
-                                                                           includingPropertiesForKeys:@[(__bridge NSString *)kCFURLNameKey]
-                                                                                              options:NSDirectoryEnumerationSkipsHiddenFiles
-                                                                                                error:&error];
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
+        NSError *error = nil;
+
+        NSURL *preHeatedCacheDirectory = [[NSBundle mainBundle] URLForResource:@"PreHeatedGraphQLCache" withExtension:nil];
+        NSArray<NSURL *> *preHeatedCacheEntries = ReadCacheEntries(preHeatedCacheDirectory, &error);
+        NSArray<NSURL *> *existingCacheEntries = error ? nil : ReadCacheEntries(_cacheDirectory, &error);
+        if (error) {
+            NSLog(@"[ARGraphQLQueryCache] Error while loading pre-heated GraphQL cache: %@", error);
+            return;
+        }
+
+        NSArray<NSString *> *existingCacheEntryKeys = [existingCacheEntries valueForKey:@"lastPathComponent"];
+
+        for (NSURL *cacheEntryURL in preHeatedCacheEntries) {
+            error = nil;
+
+            NSDictionary *cacheEntry = ReadCacheEntry(cacheEntryURL, &error);
+            if (error) {
+                NSLog(@"[ARGraphQLQueryCache] Error while loading pre-heated GraphQL cache entry `%@': %@", cacheEntryURL.path, error);
+                continue;
+            }
+
+            NSNumber *freshness = cacheEntry[@"freshness"];
+            NSParameterAssert(freshness);
+            // Skip pre-heated cache that is considered out-of-date.
+            if (!IsCacheValid([NSDate dateWithTimeIntervalSince1970:freshness.unsignedIntegerValue])) {
+                DLog(@"[ARGraphQLQueryCache] Stale pre-heated GraphQL cache entry will be ignored: %@", cacheEntryURL.path);
+                continue;
+            }
+
+            NSDictionary *queryParams = cacheEntry[@"queryParams"];
+            NSParameterAssert(queryParams);
+            NSString *queryID = queryParams[@"documentID"];
+            NSParameterAssert(queryID);
+            NSDictionary *variables = queryParams[@"variables"];
+            NSParameterAssert(variables);
+            NSString *cacheKey = CacheKey(queryID, variables);
+
+            // Skip pre-heated cache entries that already exist.
+            if ([existingCacheEntryKeys indexOfObject:cacheKey] != NSNotFound) {
+                DLog(@"[ARGraphQLQueryCache] Existing pre-heated GraphQL cache entry will be ignored: %@", cacheEntryURL.path);
+                continue;
+            }
```

Comment:
This is the actual functional change.

Reply:
Right, that makes sense. So we'll keep the first pre-heated cache the user gets and ignore new ones for upgrades.